File: server/src/Reducers/Terrain.cs
using SpacetimeDB;
using DbVector2 = pillz.server.Tables.DbVector2;
using Portal = pillz.server.Tables.Portal;
using SpawnLocation = pillz.server.Tables.SpawnLocation;

namespace pillz.server.Reducers;

public partial class Terrain
{
    [Reducer]
    public static void AddTerrainTile(ReducerContext ctx, int x, int y)
    {
        ctx.Db.Terrain.Insert(new pillz.server.Tables.Terrain
        {
            Position = new DbVector2(x, y)
        });
    }

    [Reducer]
    public static void DeleteTerrainTile(ReducerContext ctx, int x, int y)
    {
        var tile = ctx.Db.Terrain.Iter().FirstOrDefault(t => (int)t.Position.X == x && (int)t.Position.Y == y);
        ctx.Db.Terrain.Id.Delete(tile.Id);

        Log.Info($"Deleted terrain tile at ({x}, {y}) with id {tile.Id}.");
    }
    
    [Reducer]
    public static void DeleteTerrainTiles(ReducerContext ctx, float radiu)
    {
    }

    [Reducer]
    public static void GenerateTerrain(ReducerContext ctx, int seed)
    {
        var world = ctx.Db.World.Iter().FirstOrDefault();

        if (!world.IsGenerated)
        {
            GenerateTiles(ctx, seed);
            world.IsGenerated = true;
            ctx.Db.World.Id.Update(world);
        }
    }

    private static void GenerateTiles(ReducerContext ctx, int seed)
    {
        Log.Debug($"[GenerateTiles] Generating terrain with seed {seed}...");

        const int width = 300;
        const int height = 200;
        const int margin = 5;
        const int seedCount = 200;
        const int growthIterations = 4;
        const int connectThreshold = 20;

        var halfWidth = width / 2;
        var halfHeight = height / 2;

        var rng = new Random(seed);
        var  terrain = new bool[width, height];

        // STEP 1: Natural base layer with height jitter and sine waves
        var  terrainHeights = BaseLayer(width, rng,  terrain);

        // STEP 2: Place sparse random seed pixels for tunnels
        var seeds = GenerateRandomSeeds(seedCount, rng, margin, width, height,  terrain);

        // STEP 3: MST to connect all seeds with Bresenham
        ConnectSeeds(seeds,  terrain);

        // STEP 4: Tunnel drops to base only if near base
        ConnectToGround(seeds,  terrainHeights, connectThreshold,  terrain);

        // STEP 5: Growth iterations with 8-way neighbors and bias
        GrowthGeneration(growthIterations, width, height,  terrain, rng);

        // STEP 6: Write to database (centered)
        for (var x = 0; x < width; x++)
        {
            for (var y = 0; y < height; y++)
            {
                if ( terrain[x, y])
                {
                    ctx.Db.Terrain.Insert(new pillz.server.Tables.Terrain
                    {
                        Position = new DbVector2(x - halfWidth, y - halfHeight)
                    });
                }
            }
        }

        // STEP 7: Spawn and Portal locations
        GenerateSpawnLocations(ctx, width, height,  terrain, halfWidth, halfHeight);
        GeneratePortalLocations(ctx, width, height,  terrain, halfWidth, halfHeight);
    }
    
    private static void GenerateSpawnLocations(ReducerContext ctx, int width, int height, bool[,]  terrain, int halfWidth,
        int halfHeight)
    {
        for (var x = 2; x < width - 2; x++)
        {
            for (var y = 2; y < height - 3; y++) // ensure we can check y+3 safely
            {
                // This tile and the three above must be empty
                if ( terrain[x, y] ||  terrain[x, y + 1] ||  terrain[x, y + 2] ||  terrain[x, y + 3])
                    continue;

                // Ground below
                if (! terrain[x, y - 1]) continue;

                var gx = x - halfWidth;
                var gy = y - halfHeight;
                var pos = new DbVector2(gx, gy);

                ctx.Db.SpawnLocation.Insert(new SpawnLocation { Position = pos });
            }
        }
    }
    
    private static void GeneratePortalLocations(ReducerContext ctx, int width, int height, bool[,]  terrain, 
        int halfWidth, int halfHeight)
    {
        float minAllowedY = -40f;
        
        (float x, float y)? highest = null;
        (float x, float y)? lowest = null;
        (float x, float y)? leftest = null;
        (float x, float y)? rightest = null;

        
        for (var x = 1; x < width - 2; x++) // leave 1 tile margin left/right
        {
            for (var y = 2; y < height - 5; y++) // leave space above for 3-tile portal, and 2 tiles below for checks
            {
                var spaceClear =
                    ! terrain[x, y]     && ! terrain[x + 1, y] &&
                    ! terrain[x, y + 1] && ! terrain[x + 1, y + 1] &&
                    ! terrain[x, y + 2] && ! terrain[x + 1, y + 2];

                if (!spaceClear)
                    continue;

                // Ensure solid  terrain directly below either side
                var hasGroundBelow =  terrain[x, y - 1] ||  terrain[x + 1, y - 1];
                if (!hasGroundBelow)
                    continue;

                var gx = x - halfWidth;
                var gy = y - halfHeight;
                
                if (lowest == null || gy < lowest.Value.y)
                    lowest = (gx, gy);
                if (highest == null || gy > highest.Value.y) 
                    highest = (gx, gy); 
                
                if (gy >= minAllowedY)
                {
                    if (leftest == null || gx < leftest.Value.x)
                        leftest = (gx, gy);
                    if (rightest == null || gx > rightest.Value.x)
                        rightest = (gx, gy);
                }
            }
        }

        // Portals (adjust vertical placement to match visual center of portal)
        if (lowest != null && highest != null && leftest != null && rightest != null)
        {
            var portal1 = ctx.Db.Portal.Insert(new Portal
            {
                Position = new DbVector2(lowest.Value.x, lowest.Value.y)
            });

            var portal2 = ctx.Db.Portal.Insert(new Portal
            {
                Position = new DbVector2(highest.Value.x, highest.Value.y),
                ConnectedPortalId = portal1.Id
            });
            
            var portal3 = ctx.Db.Portal.Insert(new Portal
            {
                Position = new DbVector2(leftest.Value.x, leftest.Value.y),
                ConnectedPortalId = portal2.Id
            });
            
            var portal4 = ctx.Db.Portal.Insert(new Portal
            {
                Position = new DbVector2(rightest.Value.x, rightest.Value.y),
                ConnectedPortalId = portal3.Id
            });

            portal1.ConnectedPortalId = portal2.Id;
            ctx.Db.Portal.Id.Update(portal1);
            
            portal3.ConnectedPortalId = portal4.Id;
            ctx.Db.Portal.Id.Update(portal3);
        }
    }


    private static void ConnectToGround(List<(int x, int y)> seeds, int[]  terrainHeights, int connectThreshold,
        bool[,]  terrain)
    {
        foreach (var (x, y) in seeds)
        {
            var baseY =  terrainHeights[x];
            if (y - baseY <= connectThreshold)
            {
                for (var yy = y; yy >= baseY; yy--)
                     terrain[x, yy] = true;
            }
        }
    }

    private static void GrowthGeneration(int growthIterations, int width, int height, bool[,]  terrain, Random rng)
    {
        for (var i = 0; i < growthIterations; i++)
        {
            var growList = new List<(int x, int y)>();

            for (var x = 1; x < width - 1; x++)
            {
                for (var y = 1; y < height - 1; y++)
                {
                    if ( terrain[x, y]) continue;

                    var count = 0;
                    for (var dx = -1; dx <= 1; dx++)
                    {
                        for (var dy = -1; dy <= 1; dy++)
                        {
                            if (dx == 0 && dy == 0) continue;
                            if ( terrain[x + dx, y + dy]) count++;
                        }
                    }

                    // Randomness makes tunnels more varied
                    var bias = rng.Next(0, 3);

                    if (count + bias >= 5)
                    {
                        growList.Add((x, y));

                        if (count >= 6 && rng.NextDouble() < 0.75)
                        {
                            // Grow extra to add thickness
                            for (var dx = -1; dx <= 1; dx++)
                            for (var dy = -1; dy <= 1; dy++)
                                if (x + dx > 0 && y + dy > 0 && x + dx < width && y + dy < height)
                                    growList.Add((x + dx, y + dy));
                        }
                    }
                }
            }

            foreach (var (x, y) in growList)
            {
                 terrain[x, y] = true;
            }
        }
    }

    private static void ConnectSeeds(List<(int x, int y)> seeds, bool[,]  terrain)
    {
        var connected = new List<(int x, int y)> { seeds[0] };
        var remaining = new List<(int x, int y)>(seeds.Skip(1));

        while (remaining.Count > 0)
        {
            double minDist = double.MaxValue;
            (int x1, int y1) = (0, 0);
            (int x2, int y2) = (0, 0);

            foreach (var a in connected)
            {
                foreach (var b in remaining)
                {
                    double dist = Math.Pow(a.x - b.x, 2) + Math.Pow(a.y - b.y, 2);
                    if (dist < minDist)
                    {
                        minDist = dist;
                        x1 = a.x;
                        y1 = a.y;
                        x2 = b.x;
                        y2 = b.y;
                    }
                }
            }

            foreach (var (x, y) in BresenhamLine(x1, y1, x2, y2))
            {
                 terrain[x, y] = true;
            }

            connected.Add((x2, y2));
            remaining.Remove((x2, y2));
        }
    }

    private static List<(int x, int y)> GenerateRandomSeeds(int seedCount, Random rng, int margin, int width,
        int height, bool[,]  terrain)
    {
        var seeds = new List<(int x, int y)>();
        for (var i = 0; i < seedCount; i++)
        {
            var x = rng.Next(margin, width - margin);
            var y = rng.Next(15, height - margin);

            if (! terrain[x, y])
            {
                 terrain[x, y] = true;
                seeds.Add((x, y));

                // Add chunk
                for (int dx = -1; dx <= 1; dx++)
                {
                    for (int dy = -1; dy <= 1; dy++)
                    {
                        if (rng.NextDouble() < 0.8)
                             terrain[x + dx, y + dy] = true;
                    }
                }
            }
        }

        return seeds;
    }

    private static int[] BaseLayer(int width, Random rng, bool[,]  terrain)
    {
        var  terrainHeights = new int[width];
        for (var x = 0; x < width; x++)
        {
            var noise = Math.Sin((x + rng.NextDouble()) * 0.1);
            var h = 8 + (int)(noise * 4) + rng.Next(2); // more organic variation
             terrainHeights[x] = h;

            for (var y = 0; y < h; y++)
            {
                 terrain[x, y] = true;
            }
        }

        return  terrainHeights;
    }

    private static IEnumerable<(int x, int y)> BresenhamLine(int x0, int y0, int x1, int y1)
    {
        int dx = Math.Abs(x1 - x0), sx = x0 < x1 ? 1 : -1;
        int dy = -Math.Abs(y1 - y0), sy = y0 < y1 ? 1 : -1;
        int err = dx + dy;

        while (true)
        {
            yield return (x0, y0);
            if (x0 == x1 && y0 == y1) break;
            int e2 = 2 * err;
            if (e2 >= dy)
            {
                err += dy;
                x0 += sx;
            }

            if (e2 <= dx)
            {
                err += dx;
                y0 += sy;
            }
        }
    }
}

File: server/src/reducers/terrain.rs
use spacetimedb::{ReducerContext, Table};
use spacetimedb::log::{debug, info, error};
use rand::Rng;
use std::collections::HashMap;
use crate::math::dbvector2::DbVector2;
use crate::tables::{
    terrain::Terrain as TerrainTile,
    portal::Portal,
};
use crate::tables::portal::portal;
use crate::tables::terrain::{terrain};
use crate::tables::world::world;

const WIDTH: usize = 300;
const HEIGHT: usize = 200;
const MARGIN: usize = 5;
const SEED_COUNT: usize = 200;
const GROWTH_ITERATIONS: usize = 4;
const CONNECT_THRESHOLD: i32 = 20;

#[spacetimedb::reducer]
pub fn add_terrain_tile(ctx: &ReducerContext, x: i32, y: i32) -> Result<(), String> {
    ctx.db.terrain().insert(TerrainTile {
        id: 0,
        position: DbVector2::new(x as f32, y as f32),
        is_spawnable: false,
    });

    Ok(())
}

#[spacetimedb::reducer]
pub fn delete_terrain_tile(ctx: &ReducerContext, x: i32, y: i32) -> Result<(), String> {
    if let Some(tile) = ctx.db.terrain().iter()
        .find(|t| t.position.x as i32 == x && t.position.y as i32 == y)
    {
        ctx.db.terrain().id().delete(tile.id);
        info!("Deleted terrain tile at ({}, {}) with id {}.", x, y, tile.id);
    }

    Ok(())
}

#[spacetimedb::reducer]
pub fn delete_terrain_tiles(ctx: &ReducerContext, x: i32, y: i32, radius: f32) -> Result<(), String> {
    let cx = x as f32;
    let cy = y as f32;
    let r2 = radius * radius;

    let tiles_to_delete: Vec<_> = ctx.db.terrain().iter()
        .filter(|t| {
            let dx = t.position.x - cx;
            let dy = t.position.y - cy;
            dx*dx + dy*dy <= r2
        })
        .collect();

    for tile in &tiles_to_delete {
        ctx.db.terrain().id().delete(tile.id);
        info!("[DeleteTerrainTiles] Deleted terrain tile at ({}, {}) with id {}.",
              tile.position.x, tile.position.y, tile.id);
    }

    debug!("[DeleteTerrainTiles] Deleted {} tiles around ({}, {}) with radius {}.",
           tiles_to_delete.len(), x, y, radius);

    Ok(())
}

#[spacetimedb::reducer]
pub fn generate_terrain(ctx: &ReducerContext, seed: i32) -> Result<(), String> {
    // Use first World row
    let mut world = if let Some(w) = ctx.db.world().iter().next() {
        w
    } else {
        error!("World is not initialized; cannot generate terrain.");
        return Ok(())
    };

    if !world.is_generated {
        generate_tiles(ctx, seed as u64);
        world.is_generated = true;
        ctx.db.world().id().update(world);
    }

    Ok(())

}

fn generate_tiles(ctx: &ReducerContext, seed: u64) {
    debug!("[GenerateTiles] Generating terrain with seed {}...", seed);

    let half_width = (WIDTH / 2) as i32;
    let half_height = (HEIGHT / 2) as i32;

    let rng = &mut ctx.rng();

    let mut terrain = vec![vec![false; HEIGHT]; WIDTH];

    // STEP 1: Natural base layer with height jitter and sine waves
    let terrain_heights = base_layer(WIDTH, rng, &mut terrain);

    // STEP 2: Place sparse random seed pixels for tunnels
    let seeds = generate_random_seeds(SEED_COUNT, rng, MARGIN, WIDTH, HEIGHT, &mut terrain);

    // STEP 3: MST to connect all seeds with Bresenham
    connect_seeds(&seeds, &mut terrain);

    // STEP 4: Tunnel drops to base only if near base
    connect_to_ground(&seeds, &terrain_heights, CONNECT_THRESHOLD, &mut terrain);

    // STEP 5: Growth iterations with 8-way neighbors and bias
    growth_generation(GROWTH_ITERATIONS, WIDTH, HEIGHT, &mut terrain, rng);

    // STEP 6: Write to database
    let mut terrain_dict: HashMap<(i32, i32), TerrainTile> = HashMap::new();

    for x in 0..WIDTH {
        for y in 0..HEIGHT {
            if terrain[x][y] {
                let pos = DbVector2::new(
                    (x as i32 - half_width) as f32,
                    (y as i32 - half_height) as f32,
                );
                let inserted = ctx.db.terrain().insert(TerrainTile {
                    id: 0,
                    position: pos,
                    is_spawnable: false,
                });
                terrain_dict.insert((pos.x as i32, pos.y as i32), inserted);
            }
        }
    }

    // STEP 7: Spawn and Portal locations
    generate_spawn_locations(ctx, WIDTH, HEIGHT, &terrain, half_width, half_height, &mut terrain_dict);
    generate_portal_locations(ctx, WIDTH, HEIGHT, &terrain, half_width, half_height);
}

fn generate_spawn_locations(
    ctx: &ReducerContext,
    width: usize,
    height: usize,
    terrain: &[Vec<bool>],
    half_width: i32,
    half_height: i32,
    terrain_dict: &mut HashMap<(i32, i32), TerrainTile>,
) {
    for x in 2..(width - 2) {
        for y in 2..(height - 3) {
            // This tile and the three above must be empty
            if terrain[x][y] || terrain[x][y + 1] || terrain[x][y + 2] || terrain[x][y + 3] {
                continue;
            }
            // Ground below
            if !terrain[x][y - 1] {
                continue;
            }

            // world position of the GROUND tile (below empty space)
            let gx = x as i32 - half_width;
            let gy = (y as i32 - 1) - half_height;
            let key = (gx, gy);

            if let Some(mut row) = terrain_dict.get(&key).cloned() {
                row.is_spawnable = true;
                ctx.db.terrain().id().update(row);
                terrain_dict.insert(key, row);
            }
        }
    }
}

fn generate_portal_locations(
    ctx: &ReducerContext,
    width: usize,
    height: usize,
    terrain: &[Vec<bool>],
    half_width: i32,
    half_height: i32,
) {
    let min_allowed_y: f32 = -40.0;

    let mut highest: Option<(f32, f32)> = None;
    let mut lowest: Option<(f32, f32)> = None;
    let mut leftest: Option<(f32, f32)> = None;
    let mut rightest: Option<(f32, f32)> = None;

    for x in 1..(width - 2) {
        for y in 2..(height - 5) {
            let space_clear = !terrain[x][y]
                && !terrain[x + 1][y]
                && !terrain[x][y + 1]
                && !terrain[x + 1][y + 1]
                && !terrain[x][y + 2]
                && !terrain[x + 1][y + 2];

            if !space_clear {
                continue;
            }

            // Ensure solid terrain directly below either side
            let has_ground_below = terrain[x][y - 1] || terrain[x + 1][y - 1];
            if !has_ground_below {
                continue;
            }

            let gx = x as i32 - half_width;
            let gy = y as i32 - half_height;

            let gx_f = gx as f32;
            let gy_f = gy as f32;

            if lowest.map_or(true, |(_, ly)| gy_f < ly) {
                lowest = Some((gx_f, gy_f));
            }
            if highest.map_or(true, |(_, hy)| gy_f > hy) {
                highest = Some((gx_f, gy_f));
            }

            if gy_f >= min_allowed_y {
                if leftest.map_or(true, |(lx, _)| gx_f < lx) {
                    leftest = Some((gx_f, gy_f));
                }
                if rightest.map_or(true, |(rx, _)| gx_f > rx) {
                    rightest = Some((gx_f, gy_f));
                }
            }
        }
    }

    if let (Some(lowest), Some(highest), Some(leftest), Some(rightest)) =
        (lowest, highest, leftest, rightest)
    {
        let mut portal1 = ctx.db.portal().insert(Portal {
            id: 0,
            connected_portal_id: 0,
            position: DbVector2::new(lowest.0, lowest.1),
        });

        let portal2 = ctx.db.portal().insert(Portal {
            id: 0,
            connected_portal_id: portal1.id,
            position: DbVector2::new(highest.0, highest.1),
        });

        let portal3 = ctx.db.portal().insert(Portal {
            id: 0,
            connected_portal_id: portal2.id,
            position: DbVector2::new(leftest.0, leftest.1),
        });

        let portal4 = ctx.db.portal().insert(Portal {
            id: 0,
            connected_portal_id: portal3.id,
            position: DbVector2::new(rightest.0, rightest.1),
        });

        // make 1 <-> 2 and 3 <-> 4
        portal1.connected_portal_id = portal2.id;
        ctx.db.portal().id().update(portal1);

        let mut portal3_updated = portal3;
        portal3_updated.connected_portal_id = portal4.id;
        ctx.db.portal().id().update(portal3_updated);
    }
}

fn connect_to_ground(
    seeds: &[(usize, usize)],
    terrain_heights: &[i32],
    connect_threshold: i32,
    terrain: &mut [Vec<bool>],
) {
    for &(x, y) in seeds {
        let base_y = terrain_heights[x];
        let y_i = y as i32;
        if y_i - base_y <= connect_threshold {
            for yy in (base_y..=y_i).rev() {
                if yy >= 0 {
                    terrain[x][yy as usize] = true;
                }
            }
        }
    }
}

fn growth_generation(
    growth_iterations: usize,
    width: usize,
    height: usize,
    terrain: &mut [Vec<bool>],
    rng: &mut impl Rng,
) {
    for _ in 0..growth_iterations {
        let mut grow_list: Vec<(usize, usize)> = Vec::new();

        for x in 1..(width - 1) {
            for y in 1..(height - 1) {
                if terrain[x][y] {
                    continue;
                }

                let mut count = 0;
                for dx in -1..=1 {
                    for dy in -1..=1 {
                        if dx == 0 && dy == 0 {
                            continue;
                        }
                        let nx = (x as isize + dx) as usize;
                        let ny = (y as isize + dy) as usize;
                        if terrain[nx][ny] {
                            count += 1;
                        }
                    }
                }

                let bias: i32 = rng.gen_range(0..=2);
                if (count as i32 + bias) >= 5 {
                    grow_list.push((x, y));

                    if count >= 6 && rng.gen::<f64>() < 0.75 {
                        for dx in -1..=1 {
                            for dy in -1..=1 {
                                let nx = x as isize + dx;
                                let ny = y as isize + dy;
                                if nx > 0 && ny > 0 && nx < width as isize && ny < height as isize {
                                    grow_list.push((nx as usize, ny as usize));
                                }
                            }
                        }
                    }
                }
            }
        }

        for (x, y) in grow_list {
            terrain[x][y] = true;
        }
    }
}

fn connect_seeds(seeds: &[(usize, usize)], terrain: &mut [Vec<bool>]) {
    if seeds.is_empty() {
        return;
    }

    let mut connected: Vec<(usize, usize)> = vec![seeds[0]];
    let mut remaining: Vec<(usize, usize)> = seeds[1..].to_vec();

    while !remaining.is_empty() {
        let mut min_dist = f64::MAX;
        let mut x1 = 0i32;
        let mut y1 = 0i32;
        let mut x2 = 0i32;
        let mut y2 = 0i32;

        for &(ax, ay) in &connected {
            for &(bx, by) in &remaining {
                let dx = ax as f64 - bx as f64;
                let dy = ay as f64 - by as f64;
                let dist = dx * dx + dy * dy;
                if dist < min_dist {
                    min_dist = dist;
                    x1 = ax as i32;
                    y1 = ay as i32;
                    x2 = bx as i32;
                    y2 = by as i32;
                }
            }
        }

        for (x, y) in bresenham_line(x1, y1, x2, y2) {
            if x >= 0 && y >= 0 {
                let xu = x as usize;
                let yu = y as usize;
                if xu < terrain.len() && yu < terrain[0].len() {
                    terrain[xu][yu] = true;
                }
            }
        }

        let target = (x2 as usize, y2 as usize);
        connected.push(target);
        if let Some(pos) = remaining.iter().position(|&p| p == target) {
            remaining.remove(pos);
        }
    }
}

fn generate_random_seeds(
    seed_count: usize,
    rng: &mut impl Rng,
    margin: usize,
    width: usize,
    height: usize,
    terrain: &mut [Vec<bool>],
) -> Vec<(usize, usize)> {
    let mut seeds = Vec::with_capacity(seed_count);
    for _ in 0..seed_count {
        let x = rng.gen_range(margin..(width - margin));
        let y = rng.gen_range(15..(height - margin));

        if !terrain[x][y] {
            terrain[x][y] = true;
            seeds.push((x, y));

            // Add chunk
            for dx in -1..=1 {
                for dy in -1..=1 {
                    let nx = x as isize + dx;
                    let ny = y as isize + dy;
                    if nx >= 0 && ny >= 0 && nx < width as isize && ny < height as isize {
                        if rng.gen::<f64>() < 0.8 {
                            terrain[nx as usize][ny as usize] = true;
                        }
                    }
                }
            }
        }
    }
    seeds
}

fn base_layer(width: usize, rng: &mut impl Rng, terrain: &mut [Vec<bool>]) -> Vec<i32> {
    let mut heights = vec![0i32; width];
    for x in 0..width {
        let noise = ((x as f64 + rng.gen::<f64>()) * 0.1).sin();
        let h = 8 + (noise * 4.0) as i32 + rng.gen_range(0..=1);
        heights[x] = h.max(0);

        for y in 0..(h.max(0) as usize) {
            if y < terrain[0].len() {
                terrain[x][y] = true;
            }
        }
    }
    heights
}

fn bresenham_line(mut x0: i32, mut y0: i32, x1: i32, y1: i32) -> Vec<(i32, i32)> {
    let mut points = Vec::new();

    let dx = (x1 - x0).abs();
    let sx = if x0 < x1 { 1 } else { -1 };
    let dy = -(y1 - y0).abs();
    let sy = if y0 < y1 { 1 } else { -1 };
    let mut err = dx + dy;

    loop {
        points.push((x0, y0));
        if x0 == x1 && y0 == y1 { break; }
        let e2 = 2 * err;
        if e2 >= dy {
            err += dy;
            x0 += sx;
        }
        if e2 <= dx {
            err += dx;
            y0 += sy;
        }
    }
    points
}
